### PR TITLE
Added NPC and Object search functions

### PR DESCRIPTION
**Pull request by StickToTheScript**
Added more NPC and Object searching functions and a function to grab the main screen bounds.

### CODE REVIEW

File: Reflection/Ground.simba [hunk 1]
```diff
@@ -10,21 +10,52 @@ Example:
...
+  I: Integer;
   ResultTile: RSTile;
   Items: Array of RSTile;
+  MSBounds: TBox;
 begin
   Items := RSGroundObject.GetTileByID(ItemID);
-  if Length(Items) < 1 then
+  MSBounds := R_MainscreenBounds;
+  if High(Items) < 1 then
     Exit;
```
Comment (franher007) [inline Reflection/Ground.simba:22]: You could invert the order of these calls to sometimes save a call

File: Reflection/Ground.simba [hunk 2]
```diff
@@ -10,21 +10,52 @@ Example:
...
 
-  for I := 0 to Length(Items) - 1 do
+  for I := 0 to High(Items) do
   begin
-    GetClientDimensions(W, H);
     ResultTile := RSTile(Items[i]).ToLocal.TileToMS;
-    if (not ((ResultTile.X < 0) or (ResultTile.X > W) or (ResultTile.Y < 0) or (ResultTile.Y > H))) then
+    if (not ((ResultTile.X > MSBounds.X1) and (ResultTile.X < MSBounds.X2) and (ResultTile.Y > MSBounds.Y1) and (ResultTile.Y < MSBounds.Y2))) then
```
Comment (franher007) [inline Reflection/Ground.simba]: Something i use in my scripts:

```
function TPoint.InBox(Box: TBox): Boolean; constref;
begin
  Result := PointInBox(Self, Box);
end; 
```
Then you can just ask ResultTile.InBox(MSBounds)

File: Reflection/Ground.simba [hunk 3]
```diff
@@ -10,21 +10,52 @@ Example:
...
+  MSBounds: TBox;
+begin
+  Items := RSGroundObject.GetTileByID(ItemID);
+  if High(Items) = -1 then
+    Exit;
+  MSBounds := R_MainscreenBounds;
+  for I := 0 to High(Items) do
+  begin
+    ResultTile := RSTile(Items[i]).ToLocal.TileToMS;
+    if (not ((ResultTile.X > MSBounds.X1) and (ResultTile.X < MSBounds.X2) and (ResultTile.Y > MSBounds.Y1) and (ResultTile.Y < MSBounds.Y2))) then
```
Comment (franher007) [inline Reflection/Ground.simba]: Same suggestion as above :)

File: Reflection/NPC.simba [hunk 4]
```diff
@@ -201,3 +222,234 @@ begin
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Returns the Pixel Coordinates of the given NPC if they are on screen
+  as a sorted array from the middle of the screen.
+
+Example:
+
+   writeln(R_NPCGetMSCoordsSort('Chicken'));
+*)
+Function R_NPCGetMSCoordsSort(Name: String): TPointArray;
```
Comment (franher007) [inline Reflection/NPC.simba]: I'm thinking here you could do something like:

```
Function R_NPCGetMSCoordsSort(Name: String): TPointArray;
var
  Tiles : TPointArray;
begin
  Tiles := R_NPCGetMSCoords(Name);
  Tiles.sort(R_MainscreenBounds.Middle);
  Result := Tiles;
end;
```

File: Reflection/NPC.simba [hunk 5]
```diff
@@ -201,3 +222,234 @@ begin
...
+R_NPCGetMSCoordsClosest
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Returns the Pixel Coordinates of the closest given NPC if they are on screen.
+
+Example:
+
+   writeln(R_NPCGetMSCoordsClosest('Chicken'));
+*)
+Function R_NPCGetMSCoordsClosest(Name: String): TPoint;
```
Comment (franher007) [inline Reflection/NPC.simba]: Similar suggestion here

File: Reflection/NPC.simba (hunk 4 above)
Comment (JGBMichalski) [reply Reflection/NPC.simba:398]: Haha. I implemented that but must've accidentally written over my changes.

Comment (StickToTheScript) [reply Reflection/NPC.simba]: I had that in there but must've accidentally written over my previous save. Adding it back in.

File: Reflection/Ground.simba (hunk 1 above)
Comment (StickToTheScript) [reply Reflection/Ground.simba:22]: Fixed! Thanks for that find!

File: Reflection/Ground.simba (hunk 2 above)
Comment (StickToTheScript) [reply Reflection/Ground.simba]: I'm surprised that I forgot about that function.. Thanks!

**Comment by Brandon-T** [comment]
For a lot of these NPC functions, it might be better to just make a:

```pascal
type TNPC = record
    name: String;
    tile: RSTile;
    combat_level: Int32;
    etc..
end;
```

and just have a function like `GetNPCs(ID: Int32): Array of TNPC` and `GetAllNPCs: Array of TNPC`

That way, you don't have so much repeated code? Example: https://github.com/SRL/SRL-Reflection-2/blob/master/lib/animable/characters/NPCs.simba#L113

**Comment by StickToTheScript** [comment]
> For a lot of these NPC functions, it might be better to just make a:
> 
> ```pascal
> type TNPC = record
>     name: String;
>     tile: RSTile;
>     combat_level: Int32;
>     etc..
> end;
> ```
> 
> and just have a function like `GetNPCs(ID: Int32): Array of TNPC` and `GetAllNPCs: Array of TNPC`
> 
> That way, you don't have so much repeated code? Example: https://github.com/SRL/SRL-Reflection-2/blob/master/lib/animable/characters/NPCs.simba#L113

Thoughts on something like the new update?

I left in the isVisible functions and reworked the R_NPCGetMSCoordsSort.

**Pull request closed without merging**